File: src/sql/libqsqlquery.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const qsqlquery_enums = enums;
const qtsqlglobal_enums = @import("libqtsqlglobal.zig").enums;
const std = @import("std");

/// https://doc.qt.io/qt-6/qsqlquery.html
pub const qsqlquery = struct {
    /// New constructs a new QSqlQuery object.
    ///
    /// ``` r: QtC.QSqlResult ```
    pub fn New(r: ?*anyopaque) QtC.QSqlQuery {
        return qtc.QSqlQuery_new(@ptrCast(r));
    }

    /// New2 constructs a new QSqlQuery object.
    ///
    ///
    pub fn New2() QtC.QSqlQuery {
        return qtc.QSqlQuery_new2();
    }

    /// New3 constructs a new QSqlQuery object.
    ///
    /// ``` db: QtC.QSqlDatabase ```
    pub fn New3(db: ?*anyopaque) QtC.QSqlQuery {
        return qtc.QSqlQuery_new3(@ptrCast(db));
    }

    /// New4 constructs a new QSqlQuery object.
    ///
    /// ``` other: QtC.QSqlQuery ```
    pub fn New4(other: ?*anyopaque) QtC.QSqlQuery {
        return qtc.QSqlQuery_new4(@ptrCast(other));
    }

    /// New5 constructs a new QSqlQuery object.
    ///
    /// ``` query: []const u8 ```
    pub fn New5(query: []const u8) QtC.QSqlQuery {
        const query_str = qtc.libqt_string{
            .len = query.len,
            .data = query.ptr,
        };

        return qtc.QSqlQuery_new5(query_str);
    }

    /// New6 constructs a new QSqlQuery object.
    ///
    /// ``` query: []const u8, db: QtC.QSqlDatabase ```
    pub fn New6(query: []const u8, db: ?*anyopaque) QtC.QSqlQuery {
        const query_str = qtc.libqt_string{
            .len = query.len,
            .data = query.ptr,
        };

        return qtc.QSqlQuery_new6(query_str, @ptrCast(db));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#operator-eq)
    ///
    /// ``` self: QtC.QSqlQuery, other: QtC.QSqlQuery ```
    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QSqlQuery_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#swap)
    ///
    /// ``` self: QtC.QSqlQuery, other: QtC.QSqlQuery ```
    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QSqlQuery_Swap(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#isValid)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn IsValid(self: ?*anyopaque) bool {
        return qtc.QSqlQuery_IsValid(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#isActive)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn IsActive(self: ?*anyopaque) bool {
        return qtc.QSqlQuery_IsActive(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#isNull)
    ///
    /// ``` self: QtC.QSqlQuery, field: i32 ```
    pub fn IsNull(self: ?*anyopaque, field: i32) bool {
        return qtc.QSqlQuery_IsNull(@ptrCast(self), @intCast(field));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#isNull)
    ///
    /// ``` self: QtC.QSqlQuery, name: []const u8 ```
    pub fn IsNull2(self: ?*anyopaque, name: []const u8) bool {
        const name_str = qtc.libqt_string{
            .len = name.len,
            .data = name.ptr,
        };
        return qtc.QSqlQuery_IsNull2(@ptrCast(self), name_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#at)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn At(self: ?*anyopaque) i32 {
        return qtc.QSqlQuery_At(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#lastQuery)
    ///
    /// ``` self: QtC.QSqlQuery, allocator: std.mem.Allocator ```
    pub fn LastQuery(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSqlQuery_LastQuery(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qsqlquery.LastQuery: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#numRowsAffected)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn NumRowsAffected(self: ?*anyopaque) i32 {
        return qtc.QSqlQuery_NumRowsAffected(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#lastError)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn LastError(self: ?*anyopaque) QtC.QSqlError {
        return qtc.QSqlQuery_LastError(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#isSelect)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn IsSelect(self: ?*anyopaque) bool {
        return qtc.QSqlQuery_IsSelect(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#size)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn Size(self: ?*anyopaque) i32 {
        return qtc.QSqlQuery_Size(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#driver)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn Driver(self: ?*anyopaque) QtC.QSqlDriver {
        return qtc.QSqlQuery_Driver(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#result)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn Result(self: ?*anyopaque) QtC.QSqlResult {
        return qtc.QSqlQuery_Result(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#isForwardOnly)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn IsForwardOnly(self: ?*anyopaque) bool {
        return qtc.QSqlQuery_IsForwardOnly(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#record)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn Record(self: ?*anyopaque) QtC.QSqlRecord {
        return qtc.QSqlQuery_Record(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#setForwardOnly)
    ///
    /// ``` self: QtC.QSqlQuery, forward: bool ```
    pub fn SetForwardOnly(self: ?*anyopaque, forward: bool) void {
        qtc.QSqlQuery_SetForwardOnly(@ptrCast(self), forward);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#exec)
    ///
    /// ``` self: QtC.QSqlQuery, query: []const u8 ```
    pub fn Exec(self: ?*anyopaque, query: []const u8) bool {
        const query_str = qtc.libqt_string{
            .len = query.len,
            .data = query.ptr,
        };
        return qtc.QSqlQuery_Exec(@ptrCast(self), query_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#value)
    ///
    /// ``` self: QtC.QSqlQuery, i: i32 ```
    pub fn Value(self: ?*anyopaque, i: i32) QtC.QVariant {
        return qtc.QSqlQuery_Value(@ptrCast(self), @intCast(i));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#value)
    ///
    /// ``` self: QtC.QSqlQuery, name: []const u8 ```
    pub fn Value2(self: ?*anyopaque, name: []const u8) QtC.QVariant {
        const name_str = qtc.libqt_string{
            .len = name.len,
            .data = name.ptr,
        };
        return qtc.QSqlQuery_Value2(@ptrCast(self), name_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#setNumericalPrecisionPolicy)
    ///
    /// ``` self: QtC.QSqlQuery, precisionPolicy: qtsqlglobal_enums.NumericalPrecisionPolicy ```
    pub fn SetNumericalPrecisionPolicy(self: ?*anyopaque, precisionPolicy: i32) void {
        qtc.QSqlQuery_SetNumericalPrecisionPolicy(@ptrCast(self), @intCast(precisionPolicy));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#numericalPrecisionPolicy)
    ///
    /// ``` self: QtC.QSqlQuery ```
    ///
    /// Returns: ``` qtsqlglobal_enums.NumericalPrecisionPolicy ```
    pub fn NumericalPrecisionPolicy(self: ?*anyopaque) i32 {
        return qtc.QSqlQuery_NumericalPrecisionPolicy(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#setPositionalBindingEnabled)
    ///
    /// ``` self: QtC.QSqlQuery, enable: bool ```
    pub fn SetPositionalBindingEnabled(self: ?*anyopaque, enable: bool) void {
        qtc.QSqlQuery_SetPositionalBindingEnabled(@ptrCast(self), enable);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#isPositionalBindingEnabled)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn IsPositionalBindingEnabled(self: ?*anyopaque) bool {
        return qtc.QSqlQuery_IsPositionalBindingEnabled(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#seek)
    ///
    /// ``` self: QtC.QSqlQuery, i: i32 ```
    pub fn Seek(self: ?*anyopaque, i: i32) bool {
        return qtc.QSqlQuery_Seek(@ptrCast(self), @intCast(i));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#next)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn Next(self: ?*anyopaque) bool {
        return qtc.QSqlQuery_Next(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#previous)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn Previous(self: ?*anyopaque) bool {
        return qtc.QSqlQuery_Previous(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#first)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn First(self: ?*anyopaque) bool {
        return qtc.QSqlQuery_First(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#last)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn Last(self: ?*anyopaque) bool {
        return qtc.QSqlQuery_Last(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#clear)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn Clear(self: ?*anyopaque) void {
        qtc.QSqlQuery_Clear(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#exec)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn Exec2(self: ?*anyopaque) bool {
        return qtc.QSqlQuery_Exec2(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#execBatch)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn ExecBatch(self: ?*anyopaque) bool {
        return qtc.QSqlQuery_ExecBatch(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#prepare)
    ///
    /// ``` self: QtC.QSqlQuery, query: []const u8 ```
    pub fn Prepare(self: ?*anyopaque, query: []const u8) bool {
        const query_str = qtc.libqt_string{
            .len = query.len,
            .data = query.ptr,
        };
        return qtc.QSqlQuery_Prepare(@ptrCast(self), query_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#bindValue)
    ///
    /// ``` self: QtC.QSqlQuery, placeholder: []const u8, val: QtC.QVariant ```
    pub fn BindValue(self: ?*anyopaque, placeholder: []const u8, val: ?*anyopaque) void {
        const placeholder_str = qtc.libqt_string{
            .len = placeholder.len,
            .data = placeholder.ptr,
        };
        qtc.QSqlQuery_BindValue(@ptrCast(self), placeholder_str, @ptrCast(val));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#bindValue)
    ///
    /// ``` self: QtC.QSqlQuery, pos: i32, val: QtC.QVariant ```
    pub fn BindValue2(self: ?*anyopaque, pos: i32, val: ?*anyopaque) void {
        qtc.QSqlQuery_BindValue2(@ptrCast(self), @intCast(pos), @ptrCast(val));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#addBindValue)
    ///
    /// ``` self: QtC.QSqlQuery, val: QtC.QVariant ```
    pub fn AddBindValue(self: ?*anyopaque, val: ?*anyopaque) void {
        qtc.QSqlQuery_AddBindValue(@ptrCast(self), @ptrCast(val));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#boundValue)
    ///
    /// ``` self: QtC.QSqlQuery, placeholder: []const u8 ```
    pub fn BoundValue(self: ?*anyopaque, placeholder: []const u8) QtC.QVariant {
        const placeholder_str = qtc.libqt_string{
            .len = placeholder.len,
            .data = placeholder.ptr,
        };
        return qtc.QSqlQuery_BoundValue(@ptrCast(self), placeholder_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#boundValue)
    ///
    /// ``` self: QtC.QSqlQuery, pos: i32 ```
    pub fn BoundValue2(self: ?*anyopaque, pos: i32) QtC.QVariant {
        return qtc.QSqlQuery_BoundValue2(@ptrCast(self), @intCast(pos));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#boundValues)
    ///
    /// ``` self: QtC.QSqlQuery, allocator: std.mem.Allocator ```
    pub fn BoundValues(self: ?*anyopaque, allocator: std.mem.Allocator) []QtC.QVariant {
        const _arr: qtc.libqt_list = qtc.QSqlQuery_BoundValues(@ptrCast(self));
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(QtC.QVariant, _arr.len) catch @panic("qsqlquery.BoundValues: Memory allocation failed");
        const _data: [*]QtC.QVariant = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#boundValueNames)
    ///
    /// ``` self: QtC.QSqlQuery, allocator: std.mem.Allocator ```
    pub fn BoundValueNames(self: ?*anyopaque, allocator: std.mem.Allocator) [][]const u8 {
        const _arr: qtc.libqt_list = qtc.QSqlQuery_BoundValueNames(@ptrCast(self));
        const _str: [*]qtc.libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |i| {
                qtc.libqt_string_free(@ptrCast(&_str[i]));
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]const u8, _arr.len) catch @panic("qsqlquery.BoundValueNames: Memory allocation failed");
        for (0.._arr.len) |i| {
            const _data = _str[i];
            const _buf = allocator.alloc(u8, _data.len) catch @panic("qsqlquery.BoundValueNames: Memory allocation failed");
            @memcpy(_buf, _data.data[0.._data.len]);
            _ret[i] = _buf;
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#boundValueName)
    ///
    /// ``` self: QtC.QSqlQuery, pos: i32, allocator: std.mem.Allocator ```
    pub fn BoundValueName(self: ?*anyopaque, pos: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSqlQuery_BoundValueName(@ptrCast(self), @intCast(pos));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qsqlquery.BoundValueName: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#executedQuery)
    ///
    /// ``` self: QtC.QSqlQuery, allocator: std.mem.Allocator ```
    pub fn ExecutedQuery(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSqlQuery_ExecutedQuery(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qsqlquery.ExecutedQuery: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#lastInsertId)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn LastInsertId(self: ?*anyopaque) QtC.QVariant {
        return qtc.QSqlQuery_LastInsertId(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#finish)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn Finish(self: ?*anyopaque) void {
        qtc.QSqlQuery_Finish(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#nextResult)
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn NextResult(self: ?*anyopaque) bool {
        return qtc.QSqlQuery_NextResult(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#seek)
    ///
    /// ``` self: QtC.QSqlQuery, i: i32, relative: bool ```
    pub fn Seek2(self: ?*anyopaque, i: i32, relative: bool) bool {
        return qtc.QSqlQuery_Seek2(@ptrCast(self), @intCast(i), relative);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#execBatch)
    ///
    /// ``` self: QtC.QSqlQuery, mode: qsqlquery_enums.BatchExecutionMode ```
    pub fn ExecBatch1(self: ?*anyopaque, mode: i32) bool {
        return qtc.QSqlQuery_ExecBatch1(@ptrCast(self), @intCast(mode));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#bindValue)
    ///
    /// ``` self: QtC.QSqlQuery, placeholder: []const u8, val: QtC.QVariant, typeVal: flag of qtsqlglobal_enums.ParamTypeFlag ```
    pub fn BindValue3(self: ?*anyopaque, placeholder: []const u8, val: ?*anyopaque, typeVal: i64) void {
        const placeholder_str = qtc.libqt_string{
            .len = placeholder.len,
            .data = placeholder.ptr,
        };
        qtc.QSqlQuery_BindValue3(@ptrCast(self), placeholder_str, @ptrCast(val), @intCast(typeVal));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#bindValue)
    ///
    /// ``` self: QtC.QSqlQuery, pos: i32, val: QtC.QVariant, typeVal: flag of qtsqlglobal_enums.ParamTypeFlag ```
    pub fn BindValue32(self: ?*anyopaque, pos: i32, val: ?*anyopaque, typeVal: i64) void {
        qtc.QSqlQuery_BindValue32(@ptrCast(self), @intCast(pos), @ptrCast(val), @intCast(typeVal));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#addBindValue)
    ///
    /// ``` self: QtC.QSqlQuery, val: QtC.QVariant, typeVal: flag of qtsqlglobal_enums.ParamTypeFlag ```
    pub fn AddBindValue2(self: ?*anyopaque, val: ?*anyopaque, typeVal: i64) void {
        qtc.QSqlQuery_AddBindValue2(@ptrCast(self), @ptrCast(val), @intCast(typeVal));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsqlquery.html#dtor.QSqlQuery)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QSqlQuery ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QSqlQuery_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qsqlquery.html#types
pub const enums = struct {
    pub const BatchExecutionMode = enum {
        pub const ValuesAsRows: i32 = 0;
        pub const ValuesAsColumns: i32 = 1;
    };
};

File: src_original/sql/libqsqlquery.cpp
#include <QAnyStringView>
#include <QList>
#include <QSqlDatabase>
#include <QSqlDriver>
#include <QSqlError>
#include <QSqlQuery>
#include <QSqlRecord>
#include <QSqlResult>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QVariant>
#include <qsqlquery.h>
#include "libqsqlquery.h"
#include "libqsqlquery.hxx"

QSqlQuery* QSqlQuery_new(QSqlResult* r) {
    return new QSqlQuery(r);
}

QSqlQuery* QSqlQuery_new2() {
    return new QSqlQuery();
}

QSqlQuery* QSqlQuery_new3(const QSqlDatabase* db) {
    return new QSqlQuery(*db);
}

QSqlQuery* QSqlQuery_new4(const QSqlQuery* other) {
    return new QSqlQuery(*other);
}

QSqlQuery* QSqlQuery_new5(const libqt_string query) {
    QString query_QString = QString::fromUtf8(query.data, query.len);
    return new QSqlQuery(query_QString);
}

QSqlQuery* QSqlQuery_new6(const libqt_string query, const QSqlDatabase* db) {
    QString query_QString = QString::fromUtf8(query.data, query.len);
    return new QSqlQuery(query_QString, *db);
}

void QSqlQuery_OperatorAssign(QSqlQuery* self, const QSqlQuery* other) {
    self->operator=(*other);
}

void QSqlQuery_Swap(QSqlQuery* self, QSqlQuery* other) {
    self->swap(*other);
}

bool QSqlQuery_IsValid(const QSqlQuery* self) {
    return self->isValid();
}

bool QSqlQuery_IsActive(const QSqlQuery* self) {
    return self->isActive();
}

bool QSqlQuery_IsNull(const QSqlQuery* self, int field) {
    return self->isNull(static_cast<int>(field));
}

bool QSqlQuery_IsNull2(const QSqlQuery* self, libqt_string name) {
    QString name_QString = QString::fromUtf8(name.data, name.len);
    return self->isNull(QAnyStringView(name_QString));
}

int QSqlQuery_At(const QSqlQuery* self) {
    return self->at();
}

libqt_string QSqlQuery_LastQuery(const QSqlQuery* self) {
    QString _ret = self->lastQuery();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

int QSqlQuery_NumRowsAffected(const QSqlQuery* self) {
    return self->numRowsAffected();
}

QSqlError* QSqlQuery_LastError(const QSqlQuery* self) {
    return new QSqlError(self->lastError());
}

bool QSqlQuery_IsSelect(const QSqlQuery* self) {
    return self->isSelect();
}

int QSqlQuery_Size(const QSqlQuery* self) {
    return self->size();
}

QSqlDriver* QSqlQuery_Driver(const QSqlQuery* self) {
    return (QSqlDriver*)self->driver();
}

QSqlResult* QSqlQuery_Result(const QSqlQuery* self) {
    return (QSqlResult*)self->result();
}

bool QSqlQuery_IsForwardOnly(const QSqlQuery* self) {
    return self->isForwardOnly();
}

QSqlRecord* QSqlQuery_Record(const QSqlQuery* self) {
    return new QSqlRecord(self->record());
}

void QSqlQuery_SetForwardOnly(QSqlQuery* self, bool forward) {
    self->setForwardOnly(forward);
}

bool QSqlQuery_Exec(QSqlQuery* self, const libqt_string query) {
    QString query_QString = QString::fromUtf8(query.data, query.len);
    return self->exec(query_QString);
}

QVariant* QSqlQuery_Value(const QSqlQuery* self, int i) {
    return new QVariant(self->value(static_cast<int>(i)));
}

QVariant* QSqlQuery_Value2(const QSqlQuery* self, libqt_string name) {
    QString name_QString = QString::fromUtf8(name.data, name.len);
    return new QVariant(self->value(QAnyStringView(name_QString)));
}

void QSqlQuery_SetNumericalPrecisionPolicy(QSqlQuery* self, int precisionPolicy) {
    self->setNumericalPrecisionPolicy(static_cast<QSql::NumericalPrecisionPolicy>(precisionPolicy));
}

int QSqlQuery_NumericalPrecisionPolicy(const QSqlQuery* self) {
    return static_cast<int>(self->numericalPrecisionPolicy());
}

void QSqlQuery_SetPositionalBindingEnabled(QSqlQuery* self, bool enable) {
    self->setPositionalBindingEnabled(enable);
}

bool QSqlQuery_IsPositionalBindingEnabled(const QSqlQuery* self) {
    return self->isPositionalBindingEnabled();
}

bool QSqlQuery_Seek(QSqlQuery* self, int i) {
    return self->seek(static_cast<int>(i));
}

bool QSqlQuery_Next(QSqlQuery* self) {
    return self->next();
}

bool QSqlQuery_Previous(QSqlQuery* self) {
    return self->previous();
}

bool QSqlQuery_First(QSqlQuery* self) {
    return self->first();
}

bool QSqlQuery_Last(QSqlQuery* self) {
    return self->last();
}

void QSqlQuery_Clear(QSqlQuery* self) {
    self->clear();
}

bool QSqlQuery_Exec2(QSqlQuery* self) {
    return self->exec();
}

bool QSqlQuery_ExecBatch(QSqlQuery* self) {
    return self->execBatch();
}

bool QSqlQuery_Prepare(QSqlQuery* self, const libqt_string query) {
    QString query_QString = QString::fromUtf8(query.data, query.len);
    return self->prepare(query_QString);
}

void QSqlQuery_BindValue(QSqlQuery* self, const libqt_string placeholder, const QVariant* val) {
    QString placeholder_QString = QString::fromUtf8(placeholder.data, placeholder.len);
    self->bindValue(placeholder_QString, *val);
}

void QSqlQuery_BindValue2(QSqlQuery* self, int pos, const QVariant* val) {
    self->bindValue(static_cast<int>(pos), *val);
}

void QSqlQuery_AddBindValue(QSqlQuery* self, const QVariant* val) {
    self->addBindValue(*val);
}

QVariant* QSqlQuery_BoundValue(const QSqlQuery* self, const libqt_string placeholder) {
    QString placeholder_QString = QString::fromUtf8(placeholder.data, placeholder.len);
    return new QVariant(self->boundValue(placeholder_QString));
}

QVariant* QSqlQuery_BoundValue2(const QSqlQuery* self, int pos) {
    return new QVariant(self->boundValue(static_cast<int>(pos)));
}

libqt_list /* of QVariant* */ QSqlQuery_BoundValues(const QSqlQuery* self) {
    QList<QVariant> _ret = self->boundValues();
    // Convert QList<> from C++ memory to manually-managed C memory
    QVariant** _arr = static_cast<QVariant**>(malloc(sizeof(QVariant*) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = new QVariant(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_list /* of libqt_string */ QSqlQuery_BoundValueNames(const QSqlQuery* self) {
    QList<QString> _ret = self->boundValueNames();
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        QString _lv_ret = _ret[i];
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _lv_b = _lv_ret.toUtf8();
        libqt_string _lv_str;
        _lv_str.len = _lv_b.length();
        _lv_str.data = static_cast<const char*>(malloc(_lv_str.len + 1));
        memcpy((void*)_lv_str.data, _lv_b.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_string QSqlQuery_BoundValueName(const QSqlQuery* self, int pos) {
    QString _ret = self->boundValueName(static_cast<int>(pos));
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string QSqlQuery_ExecutedQuery(const QSqlQuery* self) {
    QString _ret = self->executedQuery();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

QVariant* QSqlQuery_LastInsertId(const QSqlQuery* self) {
    return new QVariant(self->lastInsertId());
}

void QSqlQuery_Finish(QSqlQuery* self) {
    self->finish();
}

bool QSqlQuery_NextResult(QSqlQuery* self) {
    return self->nextResult();
}

bool QSqlQuery_Seek2(QSqlQuery* self, int i, bool relative) {
    return self->seek(static_cast<int>(i), relative);
}

bool QSqlQuery_ExecBatch1(QSqlQuery* self, int mode) {
    return self->execBatch(static_cast<QSqlQuery::BatchExecutionMode>(mode));
}

void QSqlQuery_BindValue3(QSqlQuery* self, const libqt_string placeholder, const QVariant* val, int typeVal) {
    QString placeholder_QString = QString::fromUtf8(placeholder.data, placeholder.len);
    self->bindValue(placeholder_QString, *val, static_cast<QSql::ParamType>(typeVal));
}

void QSqlQuery_BindValue32(QSqlQuery* self, int pos, const QVariant* val, int typeVal) {
    self->bindValue(static_cast<int>(pos), *val, static_cast<QSql::ParamType>(typeVal));
}

void QSqlQuery_AddBindValue2(QSqlQuery* self, const QVariant* val, int typeVal) {
    self->addBindValue(*val, static_cast<QSql::ParamType>(typeVal));
}

void QSqlQuery_Delete(QSqlQuery* self) {
    delete self;
}
